### ISSUE TITLE
API for auto-tuned parameters

### ISSUE THREAD

**Issue by jglaser** [glotzerlab/hoomd-blue#562]
## Description

Some simulations autotune slowly. It would be great to be able to restart those from a `.gsd` with optimal autotuning parameters.

## Proposed solution

Implement `[get/set]CurrentAutotunerParam()` interface on `C++` *GPU classes. Implement a property (#398) on the python side, pass to `dump.gsd()` object (by default?).

## Additional context

Autotuning in MD can take 10,000 time steps, or in HPMC (especially GPU JIT #549), up to 1-2 minutes. This change would save time at the start of the simulation and make profiling/benchmarking easier for developers.

## Developer

Yes, I can contribute. Other volunteers?

**Comment by jglaser** [comment]
It might be necessary to return and store a list / `std::vector<unsigned int>` of autotuning parameters, because a class can have multiple autotuners.

**Comment by joaander** [comment]
Classes will need to handle different numbers of or otherwise invalid autotuner parameter lists as we sometimes add or remove autotuners on the C++ side from version to version.

We also need to give users some better control over per-object autotuning. After loading in parameters, we wouldn't want to start retuning immediately. One solution to this would be to remove the automatic re-tuning feature of the autotuners and instead provide 1) an API to trigger re-tuning per-object (perhaps on the whole simulation context too as a convenience) and 2) an Updater that allows users to trigger re-tuning periodically. @b-butler do you have any ideas?

**Comment by joaander** [comment]
Here is a more complete outline of an API.

1) Autotuners start tuning by default.
2) Users can query an object to see if tuning is complete. When an object has multiple tuners, it logically ands all of the done flags together.
```
if (mc.autotuner.done)
   #do something
```
3) Autotuners don't start retuning on their own, call:
```
mc.auotuner.start_tune()
```
to start another tuning cycle. This could be scheduled via an Updater and a PeriodicTrigger.
4) Users can query and log autotuner values:
```
>>> mc.autotuner.value
[32, 64, 100028]
```
When an object has multiple tuners, it returns an array of values.
5) Users can set an autotuner value:
```
mc.autotuner.value = [64, 128, 256]
```
Once a parameter is set, autotuning is marked as done and it holds this value until `start_tune()` is called or a new value is set.
6) Minor releases may add or remove tuners for performance reasons, changing the length of the vector or the range of accepted values. Running the same simulation on the CPU, NVIDIA GPU, or AMD GPU may also result in different possible values (typically an empty list on the CPU).

Should we gracefully accept more/fewer or invalid parameters on set? Or should we error out? In either case, the set method will need to be robust. Logging and setting autotuner parameters is considered an advanced feature and not enabled by default, users will need to add extra code to achieve this. I don't see how we can standardize the semantics of the parameters well enough to make this a hard component of the object's state.

**Comment by joaander** [comment]
@jglaser has mentioned  a branch where he taught autotuners to store multidimensional values. We should wait on that branch before we implement an API to make autotuner state accessible to the user and available for restartable jobs.

**Comment by joaander** [comment]
While this API would be nice to have, it is not critical for 3.0.0. Deferring to a future 3.x release.
